### DISCUSSION TITLE
unable to parse signature algorithm

### DISCUSSION THREAD

**Discussion by paul-at** [strongswan/strongswan#2283] [Q&A] [answered]
I'm getting the following errors on loading for certificates located in `swanctl/x509` and not quite sure where to go from it:
```
loaded certificate 'C=CH, O=redacted, CN=redacted'
   unable to parse signature algorithm
OpenSSL X.509 parsing failed
```

The certificates are rsa 3072; using ed25519 had the same effect. Created using the `pki` tool according to https://docs.strongswan.org/docs/5.9/pki/pkiQuickstart.html

Both `pki --print` and `openssl x509 -text` display the certificates without errors.

I could only find a mention of this error in #157, but this issue doesn't seem to be obviously related as the certificates are created using `pki`.

Using StrongSwan 5.9.13 from Ubuntu distribution (`5.9.13-2ubuntu4`).

`loaded plugins: charon-systemd aesni aes rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs12 pgp dnskey sshkey pem openssl pkcs8 fips-prf gmp agent xcbc hmac kdf gcm drbg attr kernel-netlink resolve socket-default connmark forecast farp stroke vici updown eap-identity eap-aka eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters`

There's also a likely related error when trying to connect, which I assume is the outcome of the signature parsing error:
```
parsed IKE_AUTH request 3 [ EAP/RES/TLS ]
no usable TLS server certificate found for 'CN-redacted'
sending fatal TLS alert 'handshake failure'
```

What's interesting, there's no error using the same certificate in a previous step:
```
authentication of 'CN-redacted' (myself) with RSA signature successful
sending end entity cert "C=CH, O=redacted, CN=CN-redacted"
generating IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
```

Not quite sure where to go from this and what's causing these certificates to fail, apparently, when strongswan does something private key-related to them?

One thing I noted is that certificates on steps 1 and 3 in https://docs.strongswan.org/docs/5.9/interop/windowsUserServerStatus.html#_ike_auth_response have different CNs (`vpn.strongswan.org` and `edu.strongsec.com`). I'm not sure if it's an error in documentation or there somehow should be two different gateway certificates configured?

The private keys are stored in `private/` in PEM format, again, no issues reading them with `openssl rsa -text`.

**Comment by tobiasbrunner** [comment, marked as answer]
> I'm getting the following errors on loading for certificates located in `swanctl/x509` and not quite sure where to go from it:

Try increasing the log levels for _lib_ and _asn_ to see more about the parsing. Also, without the actual certificates, it will be difficult to help you. It's also not clear what exactly is getting parsed there and by whom.

> There's also a likely related error when trying to connect, which I assume is the outcome of the signature parsing error:

Not necessarily. We have basically no information about your keys and certificates, so what exactly is going on we can't even begin to guess.


> What's interesting, there's no error using the same certificate in a previous step:

IKE authentication and EAP-TLS authentication are quite different especially in how certificates/keys are selected.

> I'm not sure if it's an error in documentation or there somehow should be two different gateway certificates configured?

It's an option to use a different server certificate during EAP-TLS than for the VPN gateway authentication via IKE (especially if EAP-TLS is terminated on a separate AAA/RADIUS server), but in this case I think it's a mistake that happened when a real-world log was adapted for the documentation.

**Reply by paul-at** [reply]
Thank you for your help @tobiasbrunner !

Turned out in this case "unable to parse signature algorithm" occurred when StrongSwan was reading certificate request .pems placed in `x509` directory. It's really unfortunate that the default error message doesn't at least output a file name.

As you rightly predicted, removing requests from the directory and fixing this loading error didn't resolve the IKE/EAP-TLS errors. I'll mark this discussion as resolved so as not to confuse anyone who may stumble upon it and will create a new one related to IKE/EAP-TLS with the requested details if unable to solve on my own.

**Reply by tobiasbrunner** [reply]
> It's really unfortunate that the default error message doesn't at least output a file name.

The daemon does not have any file name information when it tries to parse a certificate passed by `swanctl` via vici (only the data read from the file is sent). However, `swanctl` should have printed an error message with the file name after the negative response from the daemon.

**Reply by paul-at** [reply]
It probably did, I must've been looking at `charon-systemd` output not realising I should look at the entire log including `swanctl`.
